### PR TITLE
feat: Correct stubs for TypeVars

### LINKED ISSUE #63: feat: Analyzing TypeVar Variables and their usage and generating correct stubs
### Is your feature request related to a problem?

Currently TypeVars are generated wrongly in the stubs.

### Desired solution

 The following TypeVar usages should generate correct stubs:

```py
from typing import TypeVar

# Example 1
first_type_var = TypeVar("first_type_var")
def func(a: first_type_var, b: first_type_var) -> first_type_var: ...

# Example 2
def  func[T](a: T, b: T) -> T: ...
```

The stubs should look like the following:
```py
```

To achieve this, while analyzing the code, perhaps a list of all TypeVars with their usages should be created.

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by Masara**
Closes #63

### Summary of Changes
TypeVars create correct stubs.

### CODE REVIEW

**Comment by Masara** [comment]
@lars-reimann Like I once mentioned [here](https://github.com/Safe-DS/Stub-Generator/issues/36#issuecomment-1819603549) Mypy seems to be unable to parse square brackets right after function and class names. Therefor I had to use a workaround to get the TypeVar information we need for ```def fun[T](a: T) -> T``` like functions. The parameter type and return type are seen as "special Any types" in Mypy and I used that information to assume that we've got a TypeVar. You can see the changes concerning this "workaround" here: 9c30988bbba5a7232ef9f32be4f5f219ddacfbc0

**Comment by Masara** [comment]
Also it seems like Python 3.11 can't work with the tests containing the ```def my_fun[T]``` syntax, what should I do in this case?

**Comment by lars-reimann** [comment]
> Also it seems like Python 3.11 can't work with the tests containing the `def my_fun[T]` syntax, what should I do in this case?

That syntax was introduced in Python 3.12, so that's not surprising. Until `mypy` supports [PEP 695](https://peps.python.org/pep-0695/) natively, I'm fine with adding a new issue for that syntax to the backlog. Supporting this is not an urgent feature.

**Comment by Masara** [comment]
I removed the changes and created issue #68. You can now check the PR.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub [hunk 1]
```diff
@@ -0,0 +1,17 @@
+@PythonModule("various_modules_package.type_var_module")
+package variousModulesPackage.typeVarModule
+
+@Pure
+@PythonName("type_var_func")
+fun typeVarFunc<_type_var>(
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:6]: Type parameter names should be `UpperCamelCase`. `@PythonName` is not needed for them, but the names must be replaced consistently.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	a: List<_type_var>
+) -> result1: List<_type_var>
+
+@Pure
+@PythonName("multiple_type_var")
+fun multipleTypeVar<_type_var1, _type_var2>(
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:12]: Same.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:1]: A test for a class with type parameters is missing.

**Review by lars-reimann — CHANGES_REQUESTED** [review]
Nice work, this requires only minor changes.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub
Comment (Masara) [reply tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:1]: Am I right to assume that this
```py
T = TypeVar('T')

class TypeVarClass:
    def __init__(self, items: list[T]): ...

    def type_var_class_func(self, a: T) -> T: ...
```

would look like this in the stubs
```py
class TypeVarClass<T>(
	items: List<T>
) {
	@Pure
	@PythonName("type_var_class_func")
	fun typeVarClassFunc(
		a: T
	) -> result1: T
}
```
?

Comment (lars-reimann) [reply tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:1]: Yes, although additionally the class must inherit from something like `Generic` (see [this](https://mypy.readthedocs.io/en/stable/generics.html#defining-generic-classes)).

Comment (Masara) [reply tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:1]: So it should look like this?
```py
from typing import TypeVar, Generic
T = TypeVar('T')

class TypeVarClass(Generic[T]):
    def __init__(self): ...

    def type_var_class_func(self, a: T) -> T: ...
```
to
```
class TypeVarClass<T>(
	items: List<T>
) {
	@Pure
	@PythonName("type_var_class_func")
	fun typeVarClassFunc(
		a: T
	) -> result1: T
}
```

Comment (Masara) [reply tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:1]: So it should look like this?
```py
from typing import TypeVar, Generic
T = TypeVar('T')

class TypeVarClass(Generic[T]):
    def __init__(self): ...

    def type_var_class_func(self, a: T) -> T: ...
```
to
```
class TypeVarClass<T>() {
	@Pure
	@PythonName("type_var_class_func")
	fun typeVarClassFunc(
		a: T
	) -> result1: T
}
```

Comment (lars-reimann) [reply tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/test_type_var_creation.sdsstub:1]: I'd suggest combining both of your examples, i.e. keeping the constructor parameter:

```py
from typing import TypeVar, Generic
T = TypeVar('T')

class TypeVarClass(Generic[T]):
    def __init__(self, items: list[T]): ...

    def type_var_class_func(self, a: T) -> T: ...
```

to 

```
class TypeVarClass<T>(items: List<T>) {
    @Pure
    @PythonName("type_var_class_func")
    fun typeVarClassFunc(
        a: T
    ) -> result1: T
}
```

**Comment by Masara** [comment]
@lars-reimann I added the changes you requested. It would be nice, if you also look through the new tests at ```type_var_module.py``` and maybe suggest some more cases, since I don't know how the stubs would look like and what would be legal and what not.

**Comment by Masara** [comment]
@lars-reimann I'm done here, you can check this PR now, too.

**Pull request merged**